File: android/app/src/main/java/com/purchasely/shaker/data/PremiumManager.kt
package com.purchasely.shaker.data

import android.util.Log
import com.purchasely.shaker.purchasely.PurchaselyWrapper
import io.purchasely.ext.SubscriptionsListener
import io.purchasely.models.PLYSubscriptionData
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class PremiumManager(private val wrapper: PurchaselyWrapper) {

    private val _isPremium = MutableStateFlow(false)
    val isPremium: StateFlow<Boolean> = _isPremium.asStateFlow()

    fun refreshPremiumStatus() {
        // PURCHASELY: Fetch the current user's active subscriptions to determine premium access
        // Pass false to use cached data; true forces a network refresh
        // Docs: https://docs.purchasely.com/advanced-features/subscription-status
        wrapper.userSubscriptions(false, object : SubscriptionsListener {
            override fun onSuccess(subscriptions: List<PLYSubscriptionData>) {
                val premium = subscriptions.any { subscriptionData ->
                    subscriptionData.data.subscriptionStatus?.isExpired() == false
                }
                _isPremium.value = premium
                Log.d(TAG, "[Shaker] Premium status: $premium")
            }

            override fun onFailure(error: Throwable) {
                Log.e(TAG, "[Shaker] Error checking premium: ${error.message}")
            }
        })
    }

    companion object {
        private const val TAG = "PremiumManager"
    }
}

File: ios/Shaker/Data/PremiumManager.swift
import Foundation

@MainActor
class PremiumManager: ObservableObject {

    static let shared = PremiumManager()

    @Published var isPremium = false

    private init() {}

    func refreshPremiumStatus() {
        // PURCHASELY: Fetch the current user's active subscriptions to determine premium status.
        // Call after purchase, restore, login, or app foreground to keep entitlements up to date.
        // The wrapper maps the SDK's PLYSubscription to the SDK-free SubscriptionInfo model,
        // so the data layer never imports Purchasely.
        // Docs: https://docs.purchasely.com/advanced-features/subscription-status
        PurchaselyWrapper.shared.fetchSubscriptions(
            onSuccess: { [weak self] subscriptions in
                Task { @MainActor in
                    self?.updatePremium(from: subscriptions)
                }
            },
            onError: { error in
                print("[Shaker] Error checking premium: \(error.localizedDescription)")
            }
        )
    }

    /// Compute and apply the premium status from a subscriptions list.
    /// Used both by `refreshPremiumStatus()` and by the wrapper's success_payment chain.
    func updatePremium(from subscriptions: [SubscriptionInfo]) {
        let premium = subscriptions.contains { $0.isActive }
        isPremium = premium
        print("[Shaker] Premium status: \(premium)")
    }
}
